File: y2024/day02/day02.py
from typing import List

SAMPLE = """7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9"""


def process_input() -> str:
    with open("./day02/input.txt", "r") as fp:
        content = fp.readlines()

    # content = SAMPLE.splitlines()
    for index, line in enumerate(content):
        levels = line.strip().split()
        levels = [int(i) for i in levels]
        content[index] = levels
    return content


def is_level_safe(levels: List[int]) -> bool:
    decreasing = False
    if levels[0] > levels[1]:
        decreasing = True
    else:
        decreasing = False
    index = 0
    for l in levels[1:]:
        if levels[index] == l:
            return False, index, index + 1
        if (decreasing and levels[index] < l) or (not decreasing and levels[index] > l):
            return False, index, index + 1
        if (decreasing and (levels[index] - l > 3)) or (
            not decreasing and (l - levels[index] > 3)
        ):
            return False, index, index + 1
        index += 1
    return True, -1, -1


def part01(data) -> None:
    print(f"{'*'*20}  part01  {'*'*20}")
    unsafe = 0
    safe = 0
    for levels in data:
        is_safe, _, _ = is_level_safe(levels)
        if is_safe:
            safe += 1
        else:
            unsafe += 1

    print(f"{safe=}")
    print(f"{unsafe=}")


def part02(data) -> None:
    print(f"{'*'*20}  part02  {'*'*20}")
    unsafe = 0
    safe = 0
    for levels in data:
        is_safe = False
        for idx in range(len(levels)):
            check_level = levels[:idx] + levels[idx + 1 :]
            is_safe, _, _ = is_level_safe(check_level)
            if is_safe:
                safe += 1
                break
        if not is_safe:
            unsafe += 1
    print(f"{safe=}")
    print(f"{unsafe=}")


def main() -> None:
    data = process_input()
    part01(data)
    part02(data)


if __name__ == "__main__":
    main()

File: y2024/day02/day02.go
package day02

import (
	"adventofcode/y2024/utils"
	"fmt"
	"log"
	"strconv"
	"strings"
)

func processInput(dayDir string) [][]int {
	lines := utils.ReadLines(dayDir + "/input.txt")
	parsed := [][]int{}
	for _, line := range lines {
		strnums := strings.Split(line, " ")
		nums := []int{}
		for _, n := range strnums {
			nInt, err := strconv.Atoi(n)
			if err != nil {
				log.Fatalf(
					"could not convert to integer. line: <%s> num: <%s>\n", line, n,
				)
			}
			nums = append(nums, nInt)
		}
		parsed = append(parsed, nums)
	}
	return parsed
}

func isLevelSafe(level []int) bool {
	decreasing := false
	if level[0] > level[1] {
		decreasing = true
	} else {
		decreasing = false
	}
	index := 0
	for _, l := range level[1:] {
		if level[index] == l {
			return false
		}
		if (decreasing && level[index] < l) || (!decreasing && level[index] > l) {
			return false
		}
		if (decreasing && level[index]-l > 3) || (!decreasing && l-level[index] > 3) {
			return false
		}
		index += 1
	}
	return true
}

func part01(data [][]int) {
	safe := 0
	for _, levels := range data {
		if isLevelSafe(levels) {
			safe++
		}
	}
	fmt.Println("Part 01:", safe)
}

func part02(data [][]int) {
	safe := 0
	var l []int
	for _, level := range data {
		for idx := range len(level) {
			l = append(level[:idx], level[idx+1:]...)
			if isLevelSafe(l) {
				safe += 1
				break
			}
		}
	}
	fmt.Println("Part 02:", safe)
}

func Run(dir string) {
	utils.PartPrinter("DAY 02")
	data := processInput(dir + "/day02")
	part01(data)
	part02(data)
}
